### PR TITLE
[Event Hubs Client] Track Two: Third Preview (Event Processor Load Balancing)

### PR DESCRIPTION
**Pull request by kinelski**
These changes provide automatic load balancing for Event Processors.

Important points:
- These changes also include some updates in `PartitionPump` so we can have a better exception handling. The expected behavior is still under discussion.
- Checkpoints are still not supported. This will be included in a future PR for the Third Preview.
- Solves #7375, but tests still need polishing.

Related issue: #7377

### CODE REVIEW

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/EventProcessor.cs
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/EventProcessor.cs]: nit:  Create and start.   You're not capturing ones that you've created and failed to start.   (which I think is the correct way to go)

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/EventProcessor.cs
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/EventProcessor.cs]: Any reason why you went with a `List` here?  The way that you're using it, an Enumerable would seem to suffice and give you better flexibility when calling.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/PartitionPump.cs
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/PartitionPump.cs]: No need to do this.  The type initializer will set the same value.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/PartitionPump.cs
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/PartitionPump.cs]: nit:  you may want to consider hoisting this outside the `while` loop rather than potentially allocating each tick.   _(the compiler is likely to do this as an optimization, but...)_

File: sdk/eventhub/Azure.Messaging.EventHubs/tests/Infrastructure/EventProcessorManager.cs
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/tests/Infrastructure/EventProcessorManager.cs]: I'd recommend setting a `CancellationTokenSource` at the beginning with your desired timeout and then calling `Token.ThrowIfCancellationRequested` here, rather than doing the clock management yourself.

File: sdk/eventhub/Azure.Messaging.EventHubs/tests/Infrastructure/EventProcessorManager.cs
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/tests/Infrastructure/EventProcessorManager.cs]: Not critical for this PR, but I think it would be a good idea for us to talk through a bit.   There has to be a way that we can achieve stabilization in a more deterministic way than a polling loop, and I may have some ideas after doing a ton of this type of thing for the iterator's background processing.

Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs/tests/Infrastructure/EventProcessorManager.cs]: If you choose to stick with this for now, would you be so kind as to open an issue to track revisiting?

**Review by jsquire — APPROVED** [review]
Looks good, overall.   I left a few thoughts, but there's nothing that I'd consider blocking.   Approving now, but would appreciate the chance to take another look before you merge.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/EventProcessor.cs
Comment (kinelski) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/EventProcessor.cs]: Because we are always overwriting it in `RenewOwnership` method. If we instantiated it only in the constructor, we would need to keep clearing it and setting the keys one by one on every renew operation.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/EventProcessor.cs
Comment (kinelski) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/EventProcessor.cs]: I tried so hard, but I was obviously going to forget it somewhere.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/EventProcessor.cs
Comment (kinelski) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/EventProcessor.cs]: Just my mind playing tricks on me. I'll remove it.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/EventProcessor.cs
Comment (kinelski) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/EventProcessor.cs]: We are making it a list here so we can index its elements. We need to do it when randomly picking a partition to claim.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/EventProcessor.cs
Comment (kinelski) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/EventProcessor.cs]: I saw you used `RandomGenerator` in the `EventHubScope` class. I'll update it. It's still called `s_random` in `BasicRetryPolicy`.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/EventProcessor.cs
Comment (kinelski) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/EventProcessor.cs]: No reason at all. They even have a `TimeSpan` in their names. Updating it.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/EventProcessor.cs
Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/EventProcessor.cs]: `enumerableGuy.ElementAt(5)`   is going to be more efficient.  Event if you randomly get the last element and have to walk the list, you avoid the allocations for creating the list instance.    I'd avoid doing so, personally.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/EventProcessor.cs
Comment (kinelski) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/EventProcessor.cs]: My bad. Updating it.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/EventProcessor.cs
Comment (kinelski) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/EventProcessor.cs]: No particular reason. I'll update it.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/PartitionPump.cs
Comment (kinelski) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/PartitionPump.cs]: Agreed. I'll keep the comment because I think it contains relevant information.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/EventProcessor.cs
Comment (kinelski) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/EventProcessor.cs]: Got it. I think it's reasonable.

File: sdk/eventhub/Azure.Messaging.EventHubs/tests/Infrastructure/EventProcessorManager.cs
Comment (kinelski) [reply sdk/eventhub/Azure.Messaging.EventHubs/tests/Infrastructure/EventProcessorManager.cs]: Created a new issue (#7458) so we can focus on preview 3 for now. It's not that different from the old one.

**Review by jsquire — APPROVED** [review]

**Pull request merged**
